cmd/mounter/main.go

```diff
@@ -0,0 +1,351 @@
+	}
+
+	unmountPathCommand = func(target string) ([]byte, error) {
+		return exec.Command("/usr/bin/umount", target).CombinedOutput()
+	}
+
+	fsTypeCommand = func(source string) ([]byte, error) {
+		return exec.Command("/usr/sbin/blkid", source, "-s", "TYPE", "-o", "value").CombinedOutput()
+	}
+
+	lsblkCommand = func(source string) ([]byte, error) {
+		return exec.Command("/usr/bin/lsblk", source, "-J").CombinedOutput()
+	}
+
+	createXfs = func(source string) ([]byte, error) {
+		return exec.Command("/usr/sbin/mkfs.xfs", source).CombinedOutput()
+	}
+)
+
+// DeviceInfo returns device information returned by lsblk
+type DeviceInfo struct {
+	Name       string `json:"name"`
+	Majmin     string `json:"maj:min"`
+	Rm         bool   `json:"rm"`
+	Size       string `json:"size"`
+	Readonly   bool   `json:"ro"`
+	Type       string `json:"type"`
+	Mountpoint string `json:"mountpoint"`
+}
+
+// BlockDevices is a list of DeviceInfos from the output of lsblk
+type BlockDevices struct {
+	Blockdevices []DeviceInfo `json:"blockdevices"`
+}
+
+// FindmntInfo contains parsed findmnt -J output.
+type FindmntInfo struct {
+	Target  string `json:"target"`
+	Source  string `json:"source"`
+	Fstype  string `json:"fstype"`
+	Options string `json:"options"`
+}
+
+// FileSystems is a slice of FindmntInfo, used to parse findmnt -J output
+type FileSystems struct {
+	Filesystems []FindmntInfo `json:"filesystems"`
+}
+
+func printVersion() {
+	log.Info(fmt.Sprintf("Go Version: %s", runtime.Version()))
+	log.Info(fmt.Sprintf("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH))
+}
+
+func main() {
+	var (
+		sourcePath string
+		targetPath string
+		hostPath   string
+		unmount    bool
+	)
+	flag.Set("logtostderr", "true")
+	flag.StringVar(&sourcePath, "storagePoolPath", "/source", "path the source storagePool is mounted under")
+	flag.StringVar(&targetPath, "mountPath", "/", "target path the volume should be mounted on the host")
+	flag.StringVar(&hostPath, "hostPath", "/", "path of the host in container")
+	flag.BoolVar(&unmount, "unmount", false, "set to have the target path unmounted")
+
+	// Add the zap logger flag set to the CLI. The flag set must
+	// be added before calling pflag.Parse().
+	pflag.CommandLine.AddFlagSet(zap.FlagSet())
+
+	// Add flags registered by imported packages (e.g. glog and
+	// controller-runtime)
+	pflag.CommandLine.AddGoFlagSet(flag.CommandLine)
+
+	pflag.Parse()
+
+	// Use a zap logr.Logger implementation. If none of the zap
+	// flags are configured (or if the zap flag set is not being
+	// used), this defaults to a production zap logger.
+	//
+	// The logger instantiated here can be changed to any logger
+	// implementing the logr.Logger interface. This logger will
+	// be propagated through the whole operator, generating
+	// uniform and structured logs.
+	logf.SetLogger(zap.Logger())
+
+	printVersion()
+
+	if unmount {
+		for {
+			if !unmountPath(targetPath, hostPath) {
+				time.Sleep(time.Second)
```

Comment:
So this code is more to handle the case where it takes a bit to get the deployment pod to die, and the job executes. The deployment code (below) will fix if someone unmounts. So the scenario I was worried about that the job executed before the deployment was stopped, it would re-mount what the job unmounted. So this code will then re-unmount, until the deployment is gone, and when we re-unmount and the source is the same before and after, it means nothing re-mounted, and we can exit.

Reply:
What if a pod has an hpp pvc mounted and then hpp cr is deleted which causes unmount to be called?

Reply:
Also `addFinalizer` should be changed to not do this: `obj.SetFinalizers([]string{hppFinalizer})`

Reply:
Turns out to not be a problem, because the hpp bind mounts to this mount, and if the mount is deleted the bind mount still holds a reference to the mount, and even if mount is unmount, the reference in the pod is still valid and the pod will continue to function normally. That being said, I modified the logic to ensure all the jobs finish successfully before removing the finalizer from the hpp CR.